controllers/domachine_controller.go

```diff
@@ -0,0 +1,290 @@
+	}
+
+	computesvc := computes.NewService(ctx, clusterScope)
+	droplet, err := computesvc.GetDroplet(machineScope.GetInstanceID())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if droplet == nil {
+		droplet, err = computesvc.CreateDroplet(machineScope)
+		if err != nil {
+			errs := errors.Errorf("Failed to create droplet instance for DOMachine %s/%s: %v", domachine.Namespace, domachine.Name, err)
+			machineScope.SetErrorReason(capierrors.CreateMachineError)
+			machineScope.SetErrorMessage(errs)
+			r.Recorder.Event(domachine, corev1.EventTypeWarning, "InstanceCreatingError", errs.Error())
+			return reconcile.Result{}, errs
+		}
+		r.Recorder.Eventf(domachine, corev1.EventTypeNormal, "InstanceCreated", "Created new droplet instance - %s", droplet.Name)
+	}
+
+	machineScope.SetProviderID(strconv.Itoa(droplet.ID))
+	machineScope.SetInstanceStatus(infrav1.DOResourceStatus(droplet.Status))
+
+	addrs, err := computesvc.GetDropletAddress(droplet)
+	if err != nil {
+		machineScope.SetErrorMessage(errors.New("failed to getting droplet address"))
+		return reconcile.Result{}, err
+	}
+	machineScope.SetAddresses(addrs)
+
+	// Proceed to reconcile the DOMachine state.
+	switch infrav1.DOResourceStatus(droplet.Status) {
+	case infrav1.DOResourceStatusNew:
+		machineScope.Info("Machine instance is pending", "instance-id", machineScope.GetInstanceID())
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	case infrav1.DOResourceStatusRunning:
+		machineScope.Info("Machine instance is active", "instance-id", machineScope.GetInstanceID())
+		machineScope.SetReady()
+		r.Recorder.Eventf(domachine, corev1.EventTypeNormal, "DOMachineReady", "DOMachine %s - has ready status", droplet.Name)
+		return reconcile.Result{}, nil
+	default:
+		machineScope.SetErrorReason(capierrors.UpdateMachineError)
+		machineScope.SetErrorMessage(errors.Errorf("Instance status %q is unexpected", droplet.Status))
+		return reconcile.Result{}, nil
+	}
+}
+
+func (r *DOMachineReconciler) reconcileDelete(ctx context.Context, machineScope *scope.MachineScope, clusterScope *scope.ClusterScope) (_ reconcile.Result, reterr error) {
+	machineScope.Info("Reconciling delete DOMachine")
+	domachine := machineScope.DOMachine
+
+	defer func() {
+		if reterr == nil {
+			domachine.Finalizers = util.Filter(domachine.Finalizers, infrav1.MachineFinalizer)
+		}
+	}()
+
+	computesvc := computes.NewService(ctx, clusterScope)
+	droplet, err := computesvc.GetDroplet(machineScope.GetInstanceID())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if droplet != nil {
+		if err := computesvc.DeleteDroplet(machineScope.GetInstanceID()); err != nil {
+			return reconcile.Result{}, err
+		}
+		r.Recorder.Eventf(domachine, corev1.EventTypeNormal, "InstanceDeleted", "Deleted a instance - %s", droplet.Name)
+	}
+
+	clusterScope.V(2).Info("Unable to locate droplet instance")
+	r.Recorder.Eventf(domachine, corev1.EventTypeWarning, "NoInstanceFound", "Skip deleting")
```

Comment:
Both of these statements will be run if the call to `computesvc.DeleteDroplet` is successful

Reply:
whoops, i'm wrong. already updated :grin: